### ISSUE TITLE
Fix REST client connection handling

### ISSUE THREAD

**Issue by tmetzke** [camunda/camunda#19883]
**Describe the bug**

The following integration tests don't work as expected over REST:
* `shouldActivateJobsIfBatchIsTruncated` in the [LongPollingActivateJobsTest](https://github.com/camunda/camunda/blob/main/zeebe/qa/integration-tests/src/test/java/io/camunda/zeebe/it/client/command/LongPollingActivateJobsTest.java)
* `shouldActivateJobForOpenRequest` in the [LongPollingActivateJobsTest](https://github.com/camunda/camunda/blob/main/zeebe/qa/integration-tests/src/test/java/io/camunda/zeebe/it/client/command/LongPollingActivateJobsTest.java)

**To Reproduce**

1. Add the option `true` to the test cases in the `booleans` array in the `ValueSource` annotation, basically reverting [this commit](https://github.com/camunda/camunda/pull/19333/commits/9cbadfb14bf60fcd1b695ea83684b53360d9db4f).
2. Run the tests

**Expected behavior**

The tests pass.

**Log/Stacktrace**

* The tests fail with a timeout on the Camunda client 

**Environment:**
- OS: all
- Zeebe Version: 8.6.0-alpha3
- Configuration: long polling enabled in the gateway

**Workaround**

Switch to using the gRPC protocol for job activation using long polling.

**Additional context**

This limitation is documented in our docs with https://github.com/camunda/camunda-docs/issues/4375

**Comment by tmetzke** [comment]
## Observations

Copied from https://github.com/camunda/camunda/pull/19333#issuecomment-2196947497

> @koevskinikola and I investigated the other failing IT (`shouldActivateJobForOpenRequest`):
> * If we remove `sendActivateRequestsAndClose` AFTER the test's relevant client (lines 117/118) is opened (i.e. removing it from line 120), the test passes.
> * The `sendActivateRequestsAndClose` before the test's relevant client is not affecting the result in any way.
> * It looks like the connection of our test client is interrupted if new requests come in.
> * It doesn't exactly explain why the other test case fails but could have the same root cause (long-running connections not working on the server side).

**Comment by tmetzke** [comment]
## Observations round 2

After another round of manual tests using cURL locally, I strongly assume the issue resides on the client instead of the server.
All IT scenarios are successful using cURL manually. All IT scenarios also fail using the Java client in a "production" setup, i.e. the test setup is not the root cause.

### Setups

#### Setup A
1. Start the `StandaloneBroker`.
2. Send a job activation request via cURL with a `requestTimeout` of 15 seconds.
3. Send 5 more job activation requests in parallel via cURL with a client-side timeout of 1 second.

#### Setup B

1. Setup A
2. Before the 15s are up, create a process instance with one service task that has a job type as defined by the initial activation request

#### Setup C

1. Start the `StandaloneBroker`.
2. Create 10 process instances with one service task with job type "foo" and a variable "bar" with a String value where 'x' is repeated 1,048,576 times (max message size in bytes / 4).
3. Send a job activation request via cURL for type "foo".

### Results

#### Results A

* The initial request comes back successfully with an empty list of jobs.

#### Results B

* The initial request comes back successfully with a job list containing one item.

#### Results C

* The initial request comes back successfully with a job list containing 10 items.

**Comment by koevskinikola** [comment]
ZPA mob-programming session:

We did some additional investigations on the two failing tests using Wireshark. We managed to capture the communication of both successful and failing runs for both tests. The successful runs were made by modifying the tests. The modifications are explained below.

The captures can be found on [this (internal) link](https://drive.google.com/drive/folders/1ngzdEgXipoqunsPH0H0127OnJFrCL0PY?usp=sharing).

It seems that the root cause for the two failing tests is different:

Test `shouldActivateJobsIfBatchIsTruncated`:
* Failure: we could see a `[TCP Window Full]` message that broke the transmission of the variable data, which resulted in a [RST, ACK] message from the Gateway, asking for the connection to be reset due to protocol errors.
* Success: we reduced the size of the variable, which allowed the test to pass.
* Next step: we should investigate how we can adjust the TCP window size from the client to allow for larger messages to be transmitted.

Test `shouldActivateJobForOpenRequest`:
* Failure: Once the job activation request is sent, we see a TCP [ACK] message from the client, followed by a TCP [RST, ACK]  message from the Gateway. This seems to indicate a protocol error. It's not clear why this happens.
* Success: The test performs 3 additional job activation requests after the open long-polling request. If we remove these requests, the test succeeds, i.e. data is transmitted as expected.
* Next step: investigate why a TCO [RST, ACK] may be sent in the context of Apache Http Client and a Spring Boot REST API.

**Comment by tmetzke** [comment]
Regarding `shouldActivateJobsIfBatchIsTruncated`:
* Looking at [this description](https://undertow.io/blog/2015/04/27/An-in-depth-overview-of-HTTP2.html#_flow_control), it may be that client for HTTP/2 (gRPC) can easily increase the window size. This might not be possible for HTTP/1.1 (REST).
* This might simply be a limitation of REST over HTTP/1.1 itself

**Comment by koevskinikola** [comment]
@tmetzke since TCP is a transport layer protocol, we can't control it from the application layer, i.e. an HTTP client. gRPC which is built on HTTP/2 supports multiplexing and streaming pushed the TCP window size to larger values more quickly.

I found that we can influence the TCP window size by increasing socket buffer sizes through the following Apache HttpClient 5 configuration options:

```java
final IOReactorConfig ioReactorConfig =
        IOReactorConfig.custom() // the IOReactor manages non-blocking connection
            .setSoTimeout(Timeout.ofSeconds(30)) // Overall socket timeout
            .setSndBufSize(64 * 1024) // Larger send buffer size
            .setRcvBufSize(64 * 1024) // Larger receive buffer size
            .setTcpNoDelay(true) // Disable Nagle's algorithm for lower latency
            .build();

    final HttpAsyncClientBuilder builder =
        HttpAsyncClients.custom()
            .setIOReactorConfig(ioReactorConfig)
```

However, the truncated test continues to fail. This also doesn't guarantee a stable connection, i.e. there might still be failures depending on the size of the data.

**Comment by tmetzke** [comment]
Thanks for giving this a closer look 👍 
I am still wondering why this works with cURL, though 🤔 
@megglos and maybe @npepinpe, can we document this as a "REST over HTTP/1.1" protocol limitation of the client?
I don't see much we can do about it right now. It introduces a regression when moving from gRPC to REST communication using the Java client.

**Comment by npepinpe** [comment]
AFAIK `[TCP Window Full]` is just a Wireshark thing, telling you the sender has sent a full window amount of bytes without any acks. I'm not sure why the sender would then decide to send a RST, I suspect possibly because it expects something to have gone wrong. So the question is why the client doesn't send any ACK at that point :thinking: 

I'd need to look at the dumps I guess, it's unclear to me. AFAIK this shouldn't be a limitation of HTTP/1.1, we should be able to send arbitrarily large responses - the only thing I could imagine is someone sending a RST due to a time out somewhere if the read/send doesn't finish fast enough.

**Comment by tmetzke** [comment]
Thanks, @npepinpe, that would also tie in better with my observation that cURL is capable of handling the response from the server. It indicates a specific problem with the Apache HTTP client, not the HTTP/1.1 communication. This remains a tough nut to crack 🙈

**Comment by koevskinikola** [comment]
@npepinpe the stack trace of the failure shows that a timeout is triggered. However, no configuration adjustments have impacted the timeout so far. I'll continue debugging.

**Comment by npepinpe** [comment]
So the RST likely comes from the time out, right? The question is why the client stops acknowledging - are we overwhelming it? Maybe we are?

OTOH, things I would check (and instrument with metrics):
- Connection pool settings
- Threading/Reactor settings
- Socket timeouts (not to be confused with application level timeouts, i.e. per request)

**Comment by koevskinikola** [comment]
@npepinpe so far I've tried the following configuration adjustments, but no changes in the failures:

in `HttpClientFactory`:

```java
private HttpAsyncClientBuilder defaultClientBuilder() {
  ...
      final PoolingAsyncClientConnectionManager connectionManager =
        PoolingAsyncClientConnectionManagerBuilder.create()
            .setTlsStrategy(tlsStrategy)
            .setMaxConnPerRoute(10)
            .setMaxConnTotal(20)
            .setPoolConcurrencyPolicy(PoolConcurrencyPolicy.LAX)
            .setDefaultConnectionConfig(     // <---- I also tried passing along a default connection config
                ConnectionConfig.custom()
                    .setSocketTimeout(Timeout.ofSeconds(40L))
                    .setConnectTimeout(Timeout.ofSeconds(40L))
                    .build())
            .build();

    // the IOReactorConfig is completely new. I added it to control the socket management of async connections
    final IOReactorConfig ioReactorConfig =
        IOReactorConfig.custom()
            .setSoTimeout(Timeout.ofSeconds(30)) // Overall socket timeout
            .setSndBufSize(64 * 1024) // Larger send buffer size (the idea was to influence the TCP window size)
            .setRcvBufSize(64 * 1024) // Larger receive buffer size  (the idea was to influence the TCP window size)
            .setTcpNoDelay(true) // Disable Nagle's algorithm for lower latency  (the idea was to influence the TCP window size)
            .build();

    final HttpAsyncClientBuilder builder =
        HttpAsyncClients.custom()
            .setConnectionManager(connectionManager)
            .setIOReactorConfig(ioReactorConfig)                                        // <--- the IOReacotorConfig is passed here
            .setDefaultHeaders(Collections.singletonList(acceptHeader))
            .setUserAgent("zeebe-client-java/" + VersionUtil.getVersion())
            .evictExpiredConnections()
            .setCharCodingConfig(
                CharCodingConfig.custom().setCharset(StandardCharsets.UTF_8).build())
            .evictIdleConnections(TimeValue.ofSeconds(30))
            .useSystemProperties(); // allow users to customize via system properties
...
}

private Builder defaultClientRequestConfigBuilder() {
    return RequestConfig.custom()
        .setResponseTimeout(Timeout.ofSeconds(30L))
        .setConnectionRequestTimeout(Timeout.ofSeconds(40L)) // <--- Also tried increasing the request timeout
   ... 
}
```

I think these config properties cover the first three points you listed above. I'm not sure about the last one, I'll do some research on how that can be done.

**Comment by koevskinikola** [comment]
UPDATE:
- I missed the request timeout override [here](https://github.com/camunda/camunda/blob/f16820a657862470d4833330b2780cad9fa4a952/zeebe/qa/integration-tests/src/test/java/io/camunda/zeebe/it/client/command/LongPollingActivateJobsTest.java#L45). That was overriding the socket timeout of the connections `InternalDataChannel`.
- After adjusting this timeout to 40sec, the following occurs:
   - `shouldActivateJobsIfBatchIsTruncated` - failure remains the same.
   - `shouldActivateJobForOpenRequest` - a server-side error occurs:

```java
java.util.concurrent.CompletionException: io.camunda.zeebe.client.api.command.ProblemException: Failed with code 503: 'Service Unavailable'. Details: 'class ProblemDetail {
    type: about:blank
    title: Service Unavailable
    status: 503
    detail: null
    instance: /v2/jobs/activation
}'

	at java.base/java.util.concurrent.CompletableFuture.reportJoin(CompletableFuture.java:413)
	at java.base/java.util.concurrent.CompletableFuture.join(CompletableFuture.java:2118)
	at io.camunda.zeebe.it.client.command.LongPollingActivateJobsTest.shouldActivateJobForOpenRequest(LongPollingActivateJobsTest.java:130)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:184)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:212)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:194)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:212)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:184)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:212)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:212)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:212)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:184)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:212)
	at java.base/java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:1024)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:556)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:546)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:265)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:611)
	at java.base/java.util.stream.ReferencePipeline$7$1.accept(ReferencePipeline.java:291)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:212)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:212)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:212)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1709)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:556)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:546)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:265)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:611)
	at java.base/java.util.stream.ReferencePipeline$7$1.accept(ReferencePipeline.java:291)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1709)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:556)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:546)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:265)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:611)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1597)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1597)
Caused by: io.camunda.zeebe.client.api.command.ProblemException: Failed with code 503: 'Service Unavailable'. Details: 'class ProblemDetail {
    type: about:blank
    title: Service Unavailable
    status: 503
    detail: null
    instance: /v2/jobs/activation
}'
	at io.camunda.zeebe.client.impl.http.ApiCallback.handleErrorResponse(ApiCallback.java:111)
	at io.camunda.zeebe.client.impl.http.ApiCallback.completed(ApiCallback.java:59)
	at io.camunda.zeebe.client.impl.http.ApiCallback.completed(ApiCallback.java:30)
	at org.apache.hc.core5.concurrent.BasicFuture.completed(BasicFuture.java:123)
	at org.apache.hc.core5.concurrent.ComplexFuture.completed(ComplexFuture.java:72)
	at org.apache.hc.client5.http.impl.async.InternalAbstractHttpAsyncClient$2$1.completed(InternalAbstractHttpAsyncClient.java:304)
	at org.apache.hc.core5.http.nio.support.AbstractAsyncResponseConsumer$1.completed(AbstractAsyncResponseConsumer.java:101)
	at org.apache.hc.core5.http.nio.entity.AbstractBinAsyncEntityConsumer.completed(AbstractBinAsyncEntityConsumer.java:84)
	at org.apache.hc.core5.http.nio.entity.AbstractBinDataConsumer.streamEnd(AbstractBinDataConsumer.java:81)
	at org.apache.hc.core5.http.nio.support.AbstractAsyncResponseConsumer.streamEnd(AbstractAsyncResponseConsumer.java:142)
	at org.apache.hc.client5.http.impl.async.HttpAsyncMainClientExec$1.streamEnd(HttpAsyncMainClientExec.java:251)
	at org.apache.hc.core5.http.impl.nio.ClientHttp1StreamHandler.dataEnd(ClientHttp1StreamHandler.java:270)
	at org.apache.hc.core5.http.impl.nio.ClientHttp1StreamDuplexer.dataEnd(ClientHttp1StreamDuplexer.java:366)
	at org.apache.hc.core5.http.impl.nio.AbstractHttp1StreamDuplexer.onInput(AbstractHttp1StreamDuplexer.java:334)
	at org.apache.hc.core5.http.impl.nio.AbstractHttp1IOEventHandler.inputReady(AbstractHttp1IOEventHandler.java:64)
	at org.apache.hc.core5.http.impl.nio.ClientHttp1IOEventHandler.inputReady(ClientHttp1IOEventHandler.java:41)
	at org.apache.hc.core5.reactor.InternalDataChannel.onIOEvent(InternalDataChannel.java:142)
	at org.apache.hc.core5.reactor.InternalChannel.handleIOEvent(InternalChannel.java:51)
	at org.apache.hc.core5.reactor.SingleCoreIOReactor.processEvents(SingleCoreIOReactor.java:178)
	at org.apache.hc.core5.reactor.SingleCoreIOReactor.doExecute(SingleCoreIOReactor.java:127)
	at org.apache.hc.core5.reactor.AbstractSingleCoreIOReactor.execute(AbstractSingleCoreIOReactor.java:86)
	at org.apache.hc.core5.reactor.IOReactorWorker.run(IOReactorWorker.java:44)
	at java.base/java.lang.Thread.run(Thread.java:1570)
```

**Comment by koevskinikola** [comment]
I pushed my changes to #22302

**Comment by npepinpe** [comment]
Who is producing the 503? The server is returning this? Or is the client producing this error because the response never arrives?

**Comment by koevskinikola** [comment]
The server is producing a `500` which is mapped to [an unexpected error](https://github.com/camunda/camunda/blob/870c5a442903276945fea05b0f89479940568e7e/zeebe/gateway-rest/src/main/java/io/camunda/zeebe/gateway/rest/RestErrorMapper.java#L82-L88).

When debugging locally I see the following debug message from the server side:

```
Broker is out of disk space. All client requests will be rejected
```

I also see the following client-side error during the setup phase of the tests:

```java
14:01:54.693 [] [] [httpclient-dispatch-1] ERROR org.apache.hc.client5.http.impl.async - Pool entry is not present in the set of leased entries
java.lang.IllegalStateException: Pool entry is not present in the set of leased entries
	at org.apache.hc.core5.pool.StrictConnPool.release(StrictConnPool.java:268) ~[httpcore5-5.2.5.jar:5.2.5]
	at org.apache.hc.client5.http.impl.nio.PoolingAsyncClientConnectionManager.release(PoolingAsyncClientConnectionManager.java:410) ~[httpclient5-5.3.1.jar:5.3.1]
	at org.apache.hc.client5.http.impl.async.InternalHttpAsyncExecRuntime.discardEndpoint(InternalHttpAsyncExecRuntime.java:147) ~[httpclient5-5.3.1.jar:5.3.1]
	at org.apache.hc.client5.http.impl.async.InternalHttpAsyncExecRuntime.discardEndpoint(InternalHttpAsyncExecRuntime.java:170) ~[httpclient5-5.3.1.jar:5.3.1]
	at org.apache.hc.client5.http.impl.async.InternalAbstractHttpAsyncClient$2.failed(InternalAbstractHttpAsyncClient.java:346) ~[httpclient5-5.3.1.jar:5.3.1]
	at org.apache.hc.client5.http.impl.async.AsyncRedirectExec$1.failed(AsyncRedirectExec.java:248) ~[httpclient5-5.3.1.jar:5.3.1]
	at org.apache.hc.client5.http.impl.async.AsyncHttpRequestRetryExec$1.failed(AsyncHttpRequestRetryExec.java:197) ~[httpclient5-5.3.1.jar:5.3.1]
	at org.apache.hc.client5.http.impl.async.AsyncProtocolExec$1.failed(AsyncProtocolExec.java:295) ~[httpclient5-5.3.1.jar:5.3.1]
	at org.apache.hc.client5.http.impl.async.HttpAsyncMainClientExec$1.failed(HttpAsyncMainClientExec.java:131) ~[httpclient5-5.3.1.jar:5.3.1]
	at org.apache.hc.core5.http.impl.nio.ClientHttp1StreamHandler.failed(ClientHttp1StreamHandler.java:285) ~[httpcore5-5.2.5.jar:5.2.5]
	at org.apache.hc.core5.http.impl.nio.ClientHttp1StreamDuplexer.disconnected(ClientHttp1StreamDuplexer.java:220) ~[httpcore5-5.2.5.jar:5.2.5]
	at org.apache.hc.core5.http.impl.nio.AbstractHttp1StreamDuplexer.onDisconnect(AbstractHttp1StreamDuplexer.java:408) ~[httpcore5-5.2.5.jar:5.2.5]
	at org.apache.hc.core5.http.impl.nio.AbstractHttp1IOEventHandler.disconnected(AbstractHttp1IOEventHandler.java:95) ~[httpcore5-5.2.5.jar:5.2.5]
	at org.apache.hc.core5.http.impl.nio.ClientHttp1IOEventHandler.disconnected(ClientHttp1IOEventHandler.java:41) ~[httpcore5-5.2.5.jar:5.2.5]
	at org.apache.hc.core5.reactor.InternalDataChannel.disconnected(InternalDataChannel.java:204) ~[httpcore5-5.2.5.jar:5.2.5]
	at org.apache.hc.core5.reactor.SingleCoreIOReactor.processClosedSessions(SingleCoreIOReactor.java:231) ~[httpcore5-5.2.5.jar:5.2.5]
	at org.apache.hc.core5.reactor.SingleCoreIOReactor.doTerminate(SingleCoreIOReactor.java:106) ~[httpcore5-5.2.5.jar:5.2.5]
	at org.apache.hc.core5.reactor.AbstractSingleCoreIOReactor.execute(AbstractSingleCoreIOReactor.java:93) ~[httpcore5-5.2.5.jar:5.2.5]
	at org.apache.hc.core5.reactor.IOReactorWorker.run(IOReactorWorker.java:44) ~[httpcore5-5.2.5.jar:5.2.5]
	at java.base/java.lang.Thread.run(Thread.java:1570) [?:?]
```

**Comment by koevskinikola** [comment]
Until this bug is root-caused and fixed, the limitation is documented through https://github.com/camunda/camunda-docs/issues/4375

**Comment by tmetzke** [comment]
> @tmetzke, any chance this is still coming with `8.7`?

Nobody in the API team can solve this, I think. We tried to get capacity from the ZPA engineers, @megglos mentioned it there. But so far, we don't have any. @npepinpe, do you anyone that might a good fit to look into such an issue? Maybe pairing with @koevskinikola to get up to speed?

**Comment by npepinpe** [comment]
I asked @entangled90 to take a look, I think his expertise may come in handy, and having a set of fresh eyes always helps.

**Comment by npepinpe** [comment]
![Good news everyone](https://i.giphy.com/media/v1.Y2lkPTc5MGI3NjExMmhtY2MwMXFsd2twb290Z24wZmF2azJ2NXRvcnBjem9kY3I4MGl5aSZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/8qr5b7fs7JqxrvEOzH/giphy.gif)

So @entangled90 found the issue, and will post an update + PR for it. We'll focus on fixing the JSON/raw parts first, as its easiest, and will sync with @chillleader/connectors team on how to proceed with the documents, as there I'm not sure how large we are expecting responses to be.

**Comment by entangled90** [comment]
The cause of this very strange behaviour lies in the class `io.camunda.zeebe.client.impl.http.ApiEntityConsumer`, in particular in this method:
```java
  protected int capacityIncrement() {
    return maxCapacity - entityConsumer.getBufferedBytes();
  }
```
When the response size exceeds the maxCapacity (4MB in the test), then the returned number is 0 and the HttpClient stops receiving data silently, until a timeout is reached.

We thus have to decide if we should enforce the response size on the client or on the server:
- **SERVER** :   
  all the logic resides in the server, which is potentially more efficient: 
    - the gateway can start sending back data to the client when it realizes that it has breached the message size.
    - for grpc I *think* it's already implemented, while for REST I *think* it's not  (see `ResponseMapper.RestJobActivationResult.getJobsToDefer`)
    - we don't have to stream a large response to the client, even though the client will be unable to handle it, which will lead to having to resend all jobs to the brokers
- **CLIENT**:
we enforce a response size (potentially slightly bigger than the one on the server to account for some extra fields not taked into account by the server) and we throw an exception. If an exception is thrown from here, the caller is notified with that exception from the Future complete exceptionally. 
We must also make sure that this kind of errors are handled correctly by the server, which will have to resubmit the jobs to the broker so that other workers can fetch them.

I *personally* would proceed with solution 1. as it's probably easier to implement for REST, by doing what we are already doing for grpc. To do that, we need to:
-  add a `jobsToDefer` field to  `RestJobActivationResult` 
-  make sure that the return value of `ApiEntityConsumer.capacityIncrement` is always > 0, potentially using some configured chunk size.

I verified that if capacityIncrement is > 0, then the test passes. 
However, the client receives a response which is at least 10MB. That's what made us realize that the gateway is not enforcing the messsage size limits correctly for REST.

I noticed that the response is around 10MB with the GRPC client (like the REST client) because the size limits it's checked independently for each partition and streamed back to the client, which will accumulate all the jobs into a single collection.

We may still have to keep a limit on the client (potentially a multiple of MAX_MESSAGE_SIZE) just for safety.

**Comment by npepinpe** [comment]
:+1: I think we should definitely limit it on the gateway as well, like we do for gRPC.

I still wonder if we shouldn't add a last line of defense on the client anyway, but that it doesn't fail silently but noisily if the response is too big, since the risk is that you end up running out of memory if you just accept anything. I would defer the decision though to the ZPA team, and I assume the REST API team would have some opinion on that one.

**Comment by tmetzke** [comment]
Thanks, @entangled90, really great insights and a huge step forward 👍 
I owe you a beverage of your choosing the next (and first) time we meet 🙂

**Comment by entangled90** [comment]
:beers:

**Comment by npepinpe** [comment]
> I owe you a beverage of your choosing the next (and first) time we meet 🙂

You mean at a Camunda event where Camunda is paying? :eyes:

**Comment by megglos** [comment]
Aleks was literally asking me right now what we do about this 😆  Great job @entangled90 !

**Comment by entangled90** [comment]
Hey, I opened a PR for resolving this issue. Please take a look :-)
I did not yet add a hard limit on the client, I would that it's up to you to decide if we need to add it. 
Adding it it's just a matter of throwing an exception in the `ApiEntityConsumer` class

**Comment by entangled90** [comment]
After a detailed analysis I'm implementing only a partial fix for the moment.
#### gRPC
gRPC implementation sends to the client one message per partition with size < *max_message_size*.

The client buffers all the messages from each partition, aggregating the result:
```
  max_memory_usage =~ number_partitions x max_message_size
```

#### REST
The REST implementation was aggregating all messages from all partitions into a single JSON response.
The size of the response is proportional to `number_partitions x message_size` which can easily be over *max_message_size*, especially if the number of partitions is high.

With this PR the server sends only the commands that fit in *max_message_size* (using the BrokerResponse size (in SBE) not the JSON size). The limit on the client is removed so that the behaviour is identical between gRPC and REST

As a followup, we should try to still enforce a limit on the message size in the client. 
A possible solution is to  start using SSE for streaming the jobs from the gateway as soon as they are received from the broker in a similar way to gRPC and enforce the message size on each event that is received.

**Comment by npepinpe** [comment]
We looked very briefly into chunked responses (`StreamingResponseBody`), but this is a blocking way of doing chunked responses - meaning, you implement a `void writeTo(final OutputStream output); throws IOException` method, where writing to `OutputStream` chunks the complete response (e.g. as `application/ndjson`), but the response is completed when the method returns - meaning you'd have to fully block the caller while waiting for the broker responses, thus pinning a thread (at least until we have virtual threads with synchronized support in Java 25).

So SSE seems a little better right now, as you can do a fully asynchronous response. You could also investigate just a custom `ResponseBodyEmitter` (what `SseEmitter`, and more general WebFlux, build on top of), that might work better.

**Comment by tmetzke** [comment]
Extracted from an [internal conversation](https://camunda.slack.com/archives/CSQ2E3BT4/p1740043421826309?thread_ts=1740042805.607769&cid=CSQ2E3BT4):

> This requires two fixes:
> 1. The REST gateway has to implement the same mechanism as the gRPC gateway, where we yield jobs back to the broker if they would exceed our max message size
> 2. The client should not silently discard responses that are too big: either log, throw an exception, etc. Or we do away with the "max" message size, at the cost of possibly killing the application.

**Comment by berkaycanbc** [comment]
> The REST gateway has to implement the same mechanism as the gRPC gateway, where we yield jobs back to the broker if they would exceed our max message size

@tmetzke, @npepinpe it looks like already supported in the rest gateway, right?

https://github.com/camunda/camunda/blob/11241cb92d1c81c4129a0133d2d585903636e16c/zeebe/gateway-rest/src/main/java/io/camunda/zeebe/gateway/rest/ResponseMapper.java#L708-L724

https://github.com/camunda/camunda/blob/11241cb92d1c81c4129a0133d2d585903636e16c/zeebe/gateway/src/main/java/io/camunda/zeebe/gateway/impl/job/RoundRobinActivateJobsHandler.java#L166-L174

And the long polling mechanism uses round robin handler.

Could you please clarify once more what is needed to be changed?

**Comment by tmetzke** [comment]
To be honest, I was also wondering if this is already solved on that end. But I have been away from this topic for too long. I believe @npepinpe and @entangled90 by now have a way better understanding of the issue(s) at hand, so maybe they can chime in here with thoughts.

Maybe it's easier to jump on a call for this and clarify it, maybe also by looking at some code together. I feel that, at this point, this might be the fastest way to complete this here. There have been some back-and-forths in writing that seem to leave some question marks. So maybe a sync chat is more suited to resolve this.

**Comment by entangled90** [comment]
There was also another difference between gRPC and REST:
- in gRPC a message is sent from each partition (as elements of the stream)
- REST aggregates the messages from all partitions into a single response

So at the socket level, the two message sizes will look quite different

**Comment by berkaycanbc** [comment]
Need to drop my focus on this due to priority change.

Here are my further findings:

Currently, we query each partition and retrieve jobs maximum of 4MB (default) per partition.
* gRPC Gateway : streams each response to the client as it retrieves per partition and the resulting jobs from all partitions are aggregated in the client. Meaning that client can have `partitionCount x 4MB` memory usage.
* REST Gateway:  aggregates jobs in the gateway as it retrieves per partition. Meaning that the gateway can have `partitionCount x 4MB` memory usage.

Having a different behaviour in REST and gRPC is not ideal for sure. Also, both cases might cause `maxMessageSize x 3` memory usage either in gateway or client. I expect we stream `maxMessageSize` per request in total. E.g. retrieving jobs of size `maxMessageSize / partitionCount` per partition, so we have a total of 4MB at max as a response. This comes with a cost of missing available jobs in some partitions though.

**Comment by berkaycanbc** [comment]
I started working on this again. As a start, the test `shouldActivateJobsIfBatchIsTruncated` passes with REST enabled. I will narrow down the problem to fix `shouldActivateJobForOpenRequest ` test case.

**Comment by berkaycanbc** [comment]
I started checking the difference between REST and gRPC behaviour. I've observed that for gRPC, the cancel handler:

https://github.com/camunda/camunda/blob/4e38ee8bb953ae00956094d641f7807f1584e82a/zeebe/gateway/src/main/java/io/camunda/zeebe/gateway/impl/job/LongPollingActivateJobsHandler.java#L169-L179

is invoked, so inflight request is removed. However, for REST this invocation never happens. It is due to the invocation only happens when future is completed:

https://github.com/camunda/camunda/blob/4e38ee8bb953ae00956094d641f7807f1584e82a/zeebe/gateway-rest/src/main/java/io/camunda/zeebe/gateway/rest/controller/JobController.java#L113-L118

We need to add a way to listen client close calls to the REST without necessarily waiting for result.

**Comment by tmetzke** [comment]
> We need to add a way to listen client close calls to the REST without necessarily waiting for result.

As far as I know, you cannot listen to this on the server side, there is no bidirectional channel like for gRPC.

**Comment by berkaycanbc** [comment]
> As far as I know, you cannot listen to this on the server side, there is no bidirectional channel like for gRPC.

Indeed. There seem to be some ways to mimic that. Using `DeferredResult` or starting `AsyncContext` and adding listener to it might be options. But I am not sure if these going to work reliably between asynchronous callbacks . E.g. registered listeners can still receive the callback later than job is returned to previously closed client because its callback was not yet triggered. Another option would be periodic heartbeat checks. But still need to try them out.

**Comment by berkaycanbc** [comment]
Quick update for yesterday's work. I was able to make the test succeed by using DeferredResult and its `onTimeout` listener. The idea is basically the following:

* Provide asynchronous execution via DeferredResult

```java
  @CamundaPostMapping(path = "/activation")
  public DeferredResult<ResponseEntity<Object>> activateJobs(
      @RequestBody final JobActivationRequest activationRequest, final HttpServletRequest request) {

    return RequestMapper.toJobsActivationRequest(
            activationRequest, multiTenancyCfg.isChecksEnabled())
        .fold(
            problem -> {
              final var d = new DeferredResult<ResponseEntity<Object>>(0L);
              d.setResult(RestErrorMapper.mapProblemToResponse(problem));
              return d;
            },
            r ->
                activateJobs(
                    r, activationRequest.getWorker(), activationRequest.getRequestTimeout()));
  }
```

* Listen result `CompletableFuture` and set deferred result if any jobs returned.

```java
result.whenComplete( response -> {
  if (response.jobs.isNotEmpty()) {
    deferredResult.set(response)
  }
}
```

Means that empty results will not be set, so long polling continues until `deferredResult`'s timeout.

* Invoke cancellation handler on deferredResult timeout.

```java
final var deferredResult = new DeferredResult<ResponseEntity<Object>>(requestTimeout);
deferredResult.onTimeout(() -> responseObserver.invokeCancelationHandler())
```

As deferred result has a timeout set as the same as request timeout, any disconnection will be detected early by periodic timeout/keep-alive checks. Therefore, `onTimeout` will be triggered earlier than the actual timeout.

**Summary**
Here is my understanding why it works that way. For the test case:

https://github.com/camunda/camunda/blob/5a88e2155dac62c0d17a5c02681ded2a92c62e68/zeebe/qa/integration-tests/src/test/java/io/camunda/zeebe/it/client/command/LongPollingActivateJobsTest.java#L107-L129

I see the following exception thrown for every closed client (6 times in total):

```console
12:51:17.769 [] [{}] [http-nio-0.0.0.0-1031-exec-7] DEBUG io.camunda.zeebe.gateway.rest - null
org.springframework.web.context.request.async.AsyncRequestTimeoutException
	at org.springframework.web.context.request.async.TimeoutDeferredResultProcessingInterceptor.handleTimeout(TimeoutDeferredResultProcessingInterceptor.java:42)
	at org.springframework.web.context.request.async.DeferredResultInterceptorChain.triggerAfterTimeout(DeferredResultInterceptorChain.java:81)
	at org.springframework.web.context.request.async.WebAsyncManager.lambda$startDeferredResultProcessing$5(WebAsyncManager.java:434)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
	at org.springframework.web.context.request.async.StandardServletAsyncWebRequest.onTimeout(StandardServletAsyncWebRequest.java:186)
	at org.apache.catalina.core.AsyncListenerWrapper.fireOnTimeout(AsyncListenerWrapper.java:44)
	at org.apache.catalina.core.AsyncContextImpl.timeout(AsyncContextImpl.java:136)
	at org.apache.catalina.connector.CoyoteAdapter.asyncDispatch(CoyoteAdapter.java:135)
	at org.apache.coyote.AbstractProcessor.dispatch(AbstractProcessor.java:243)
	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:57)
	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:903)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1769)
	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:52)
	at org.apache.tomcat.util.threads.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1189)
	at org.apache.tomcat.util.threads.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:658)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:63)
	at java.base/java.lang.Thread.run(Thread.java:1583)
```

With `DeferredResult` Spring’s `WebAsyncManager` wires interceptors and a `StandardServletAsyncWebRequest` that converts those async events (e.g. disconnection) into `deferredResult.onTimeout/onCompletion` callbacks. Then, triggers the`AsyncRequestTimeoutException` after the callback execution.

**Next**
I would like to try if I can achieve a similar behaviour with custom interceptors without using `DeferredResult` or at least a simpler version of it.

**Comment by megglos** [comment]
@berkaycanbc we just stumbled over the [docs that mention this limitation](https://github.com/camunda/camunda-docs/issues/4375), please make sure we remove it once fixed.

**Comment by berkaycanbc** [comment]
### Update

I worked on verifying previous findings. I discovered that the timeout was actually triggered on actual timeout and **not when http connection is closed**. Verified that the client throws `ConnectionClosedException` however none of the options I tried was able to catch the connection close. Lastly, we've paired with @entangled90 and changed Camunda Client http close from `GRACEFUL` to `IMMEDIATE` but still `onError` was not triggered at all.

Below I will document the different callbacks I've tried.

#### Test Case

```java
  @ParameterizedTest
  @ValueSource(booleans = {true})
  public void shouldActivateJobForOpenRequest(final boolean useRest, final TestInfo testInfo)
      throws InterruptedException {
    // given
    final String jobType = "job-" + testInfo.getDisplayName();

    sendActivateRequestsAndClose(useRest, jobType, "first");

    final var activateJobsResponse =
        getCommand(client, useRest).jobType(jobType).maxJobsToActivate(5).workerName("open").send();

    sendActivateRequestsAndClose(useRest, jobType, "second");

    // when
    resourcesHelper.createSingleJob(jobType);

    // then
    final var jobs = activateJobsResponse.join().getJobs();
    assertThat(jobs).hasSize(1).extracting(ActivatedJob::getWorker).contains("open");
  }

  private void sendActivateRequestsAndClose(
      final boolean useRest, final String jobType, final String phase) throws InterruptedException {
    for (int i = 0; i < 3; i++) {
      final CamundaClient tempClient = zeebe.newClientBuilder().build();

      final String workerName = "closed-" + phase + "-" + i;
      final CamundaFuture<ActivateJobsResponse> send =
          getCommand(tempClient, useRest)
              .jobType(jobType)
              .maxJobsToActivate(5)
              .workerName(workerName)
              .send();

      send.whenComplete(
          (response, cause) ->
              System.out.println(
                  "Activate request "
                      + (cause == null ? "completed" : "failed")
                      + ": "
                      + (response != null ? response.getJobs().size() + " jobs" : cause)
                      + ". workerName: "
                      + workerName));

      Thread.sleep(100);
      tempClient.close();
    }
  }
```

#### DeferredResult

```java
  @CamundaPostMapping(path = "/activation")
  public DeferredResult<ResponseEntity<Object>> activateJobs(
      @RequestBody final JobActivationRequest activationRequest) {
    final DeferredResult<ResponseEntity<Object>> deferred = new DeferredResult<>(60_000L);

    deferred.onError(
        t -> {
          System.out.println("Deferred error");
        });

    deferred.onCompletion(
        () -> {
          System.out.println("Deferred complete");
        });

    deferred.onTimeout(
        () -> {
          System.out.println("Deferred timeout");
        });

    return deferred;
  }
```

`onComplete` and `onTimeout` is called randomly for different job activation requests per test run. Because test times out already before executing callbacks. `onError` is never triggered.
 
#### AsyncListener

Added an interceptor that captures the started asynchronous context (it is not yet started in controller if you call `getAsyncContext()`). However, the result is the same, only `onComplete` is triggered and `onError` was never triggered.

```java
public final class LongPollAsyncInterceptor implements AsyncHandlerInterceptor {

  @Override
  public void afterConcurrentHandlingStarted(
      final HttpServletRequest request, final HttpServletResponse response, final Object handler) {

    // Called right after Spring starts async for controllers returning
    // CompletableFuture/Callable/etc.
    final AsyncContext async = request.getAsyncContext();
    if (async == null) {
      throw new IllegalStateException("AsyncContext is null");
    }

    async.addListener(
        new AsyncListener() {
          @Override
          public void onComplete(final AsyncEvent event) throws IOException {
            System.out.println("Async complete");
          }

          @Override
          public void onTimeout(final AsyncEvent event) throws IOException {
            System.out.println("Async timeout");
          }

          @Override
          public void onError(final AsyncEvent event) throws IOException {
            System.out.println("Async error");
          }

          @Override
          public void onStartAsync(final AsyncEvent event) throws IOException {
            System.out.println("Async start");
          }
        });
  }
}

// register the interceptor
@Configuration
public class AsyncConfig implements WebMvcConfigurer {
  @Override
  public void addInterceptors(final InterceptorRegistry registry) {
    registry.addInterceptor(new LongPollAsyncInterceptor());
  }
}
```

#### ResponseEmitter

```java
  @CamundaPostMapping(path = "/activation")
  public ResponseBodyEmitter activateJobs(
      @RequestBody final JobActivationRequest activationRequest) {
    return RequestMapper.toJobsActivationRequest(
            activationRequest, multiTenancyCfg.isChecksEnabled())
        .fold(
            err -> {
              final var emitter = new ResponseBodyEmitter();
              try {
                emitter.send(RestErrorMapper.mapProblemToResponse(err));
              } catch (final Exception e) {
                // ignore
              } finally {
                emitter.complete();
              }
              return emitter;
            },
            this::activateJobs);
  }

  private ResponseBodyEmitter activateJobs(final ActivateJobsRequest activationRequest) {
    final var result = new CompletableFuture<ResponseEntity<Object>>();
    final var responseObserver = responseObserverProvider.apply(result);
    final var responseEmitter = new ResponseBodyEmitter(activationRequest.requestTimeout());

    result.whenComplete(
        (r, e) -> {
          if (e == null) {
            try {
              System.out.println("ResponseEmitter sending response");
              responseEmitter.send(r);
              responseEmitter.complete();
            } catch (final IOException ex) {
              System.out.println("ResponseEmitter send error");
              responseObserver.invokeCancelationHandler();
            }
          } else {
            try {
              System.out.println("ResponseEmitter received error");
              responseEmitter.send(mapErrorToResponse(e));
            } catch (final IOException ex) {
              System.out.println("ResponseEmitter send error");
              responseObserver.invokeCancelationHandler();
            }
          }
        });

    responseEmitter.onError(
        err -> {
          System.out.println("ResponseEmitter error");
          responseObserver.invokeCancelationHandler();
        });

    responseEmitter.onTimeout(
        () -> {
          System.out.println("ResponseEmitter timeout");
          responseObserver.invokeCancelationHandler();
        });
    responseEmitter.onError(err -> System.out.println("ResponseEmitter error"));

    jobServices
        .withAuthentication(authenticationProvider.getCamundaAuthentication())
        .activateJobs(activationRequest, responseObserver, responseObserver::setCancelationHandler);
    return responseEmitter;
  }
```
The result was the same with using ResponseEmitter. `onError` was never triggered.

**Comment by berkaycanbc** [comment]
Well, `SseEmitter` seems to be triggered on connection close when `HttpClient.close` is called with `IMMEDIATE` mode. There were more adjustments I need to implement to make it work with SseEmitter. Please check the PR #38296 description for more details.

**Comment by aleksander-dytko** [comment]
@berkaycanbc are we planning to fix this bug for the 8.8.0 release?

**Comment by berkaycanbc** [comment]
@aleksander-dytko we do not have enough time to fix this until 8.8.0 as the freeze is on Monday. The resolution requires a broad time investigation. Even this has a potential to become a feature. See my comment https://github.com/camunda/camunda/pull/38296#issuecomment-3308894625.

**What is the cost of not fixing it?**

Jobs might be returned to closed clients but will be available for activation by another worker after job activation timeout set by the client is elapsed. Additionally, frequent client closes is not expected in job handling. Client will usually be closed on client application shutdown etc.

**How to move forward**
As we targeted this to 8.8, we can consider investing time on the fix within the next patch `8.8.1`. At least, we can decide on the scope of it. WDYT?

**Comment by aleksander-dytko** [comment]
@berkaycanbc thank you for the status. 

As the severity is defined as high, I agree we should fix this in the next patch release. 

However, as part of 8.8, I would propose that we update [this documentation](https://docs.camunda.io/docs/next/apis-tools/java-client/job-worker/#how-job-workers-work) to clearly explain what the limitation/risk is and how to avoid it. At the moment, this is not clear and might be scary for customers, especially that they will look at this page as part of the Zeebe Client -> Camunda Client migration. 

CC: @megglos

**Comment by berkaycanbc** [comment]
@aleksander-dytko @megglos Created PR https://github.com/camunda/camunda-docs/pull/6799. I will be off for the upcoming weeks, so feel free to commit to the PR if needed.

**Comment by eppdot** [comment]
@marcosgvieira @berkaycanbc this is still assigned to team alpha, thus I assume you will pick this up again, right?

(cleaning up the finalize api epic after 8.8 has been successfully released)

**Comment by berkaycanbc** [comment]
@eppdot I don't think I can pick this this up anytime soon due to `8.9` P0 work. @marcosgvieira let me know if need alignment.

**Comment by marcosgvieira** [comment]
🔄 **Status updated automatically (Core Features)**

**From:** Ready  
**To:** Backlog  

**Reason:** The issue does not have an assignee.

**Comment by eppdot** [comment]
@berkaycanbc  would you be able to pick this up again? Would be awesome to have it fixed for 8.9 (cc @marcosgvieira)

_(sidenote: I am updating the [quality board](https://github.com/orgs/camunda/projects/187/views/20) for 8.9 delivery)_

**Comment by berkaycanbc** [comment]
> [@berkaycanbc](https://github.com/berkaycanbc) would you be able to pick this up again? Would be awesome to have it fixed for 8.9 (cc [@marcosgvieira](https://github.com/marcosgvieira))
> 
> _(sidenote: I am updating the [quality board](https://github.com/orgs/camunda/projects/187/views/20) for 8.9 delivery)_

@eppdot It is hard to tell at the moment for two reasons:

1) I am still working on additional small feature work. Then, I am expected to help another epic (might not happen).
2) As seen from previous comments, we couldn't come up with a simple workaround of the problem so far. Eventually, we might choose to go for a long term solution which will require more time to solve.

If we think this has more priority over others, I suggest bringing this to whole core features team's radar.

**Comment by multani** [comment]
FYI, I was analyzing the ratio of HTTP errors vs. total requests on C8 clusters in SaaS.

I also noticed several clusters with a constant >= 5% to 15% errors on the Zeebe REST API, specifically 503s on the `/v2/jobs/activation` endpoint. So far, it seems to impact only 5 x C8 clusters (I can provide ID's privately), most of them seem to be running on "low" hardware package.

The number of impacted clusters is not very high, but this constant (comparatively) significant % of errors prevents having a reliable way of monitoring the customer's experience for these clusters.

**Comment by marcosgvieira** [comment]
🔄 **Status updated automatically (Core Features)**

**From:** Backlog  
**To:** Ready  

**Reason:** The issue already has an assignee.

**Comment by marcosgvieira** [comment]
🔄 **Iteration updated automatically (Core Features)**

**From:** Iteration 8  
**To:** Mar M2 (8.9)  

**Reason:** The issue is In Progress, so it should be part of the current iteration.

**Comment by marcosgvieira** [comment]
🔄 **Iteration updated automatically (Core Features)**

**From:** Mar M2 (8.9)  
**To:** Iteration 8  

**Reason:** The issue is In Progress, so it should be part of the current iteration.

**Comment by megglos** [comment]
We hit this from a different direction in a load test and opened #61321 for a second symptom of the same gateway behaviour: a REST worker with job streaming enabled never recovers jobs that were yielded after a failed job push, without any client being closed.

The cause is the one described here. A REST request whose client is gone still looks open to the gateway, so `getNextOpenPendingRequest` cannot skip it. It gets woken in arrival order, the job is activated into a response nobody reads, and the worker waiting behind it never gets its turn.

We mention it because the discussion here assumes the impact is limited to setups that close clients often. That is not the only case — a normal, long-running worker is affected too.

#61321 has the measurements, and a reproduction that runs locally in about 16 seconds using the REST case of `LongPollingActivateJobsTest.shouldActivateJobForOpenRequest`, which is currently switched off with a TODO pointing at this issue.
